crates/edr_provider/src/data.rs

```diff
@@ -100,60 +105,87 @@ pub struct ProviderData {
 
         let BlockchainAndState {
             blockchain,
             fork_metadata,
             state,
             irregular_state,
-        } = create_blockchain_and_state(runtime, config, genesis_accounts).await?;
+        } = create_blockchain_and_state(runtime_handle.clone(), &config, genesis_accounts).await?;
 
         let prev_randao_generator = RandomHashGenerator::with_seed("randomMixHashSeed");
 
+        let allow_blocks_with_same_timestamp = config.allow_blocks_with_same_timestamp;
+        let allow_unlimited_contract_size = config.allow_unlimited_contract_size;
+        let beneficiary = config.coinbase;
+        let block_gas_limit = config.block_gas_limit;
+        let block_time_offset_seconds = block_time_offset_seconds(&config)?;
+        let is_auto_mining = config.mining.auto_mine;
+
         Ok(Self {
+            runtime_handle,
+            initial_config: Some(config),
+
             blockchain,
             state,
             irregular_state,
-            mem_pool: MemPool::new(config.block_gas_limit),
-            network_id: config.network_id,
-            beneficiary: config.coinbase,
+            mem_pool: MemPool::new(block_gas_limit),
+            beneficiary,
             // TODO: Add config option (https://github.com/NomicFoundation/edr/issues/111)
             // Matches Hardhat default
             min_gas_price: U256::ZERO,
             prev_randao_generator,
-            block_time_offset_seconds: block_time_offset_seconds(config)?,
+            block_time_offset_seconds,
             fork_metadata,
             instance_id: B256::random(),
-            is_auto_mining: config.mining.auto_mine,
+            is_auto_mining,
             next_block_base_fee_per_gas: None,
             next_block_timestamp: None,
             // Start with 1 to mimic Ganache
             next_snapshot_id: 1,
             snapshots: BTreeMap::new(),
-            allow_blocks_with_same_timestamp: config.allow_blocks_with_same_timestamp,
-            allow_unlimited_contract_size: config.allow_unlimited_contract_size,
+            allow_blocks_with_same_timestamp,
+            allow_unlimited_contract_size,
             local_accounts,
             filters: HashMap::default(),
             last_filter_id: U256::ZERO,
             logger: Logger::new(false),
             impersonated_accounts: HashSet::new(),
-            callbacks,
+            callbacks: Some(callbacks),
         })
     }
 
+    pub async fn reset(&mut self, fork_config: Option<ForkConfig>) -> Result<(), CreationError> {
+        let mut config = std::mem::take(&mut self.initial_config).expect("initial_config exists");
+        config.fork = fork_config;
+
+        let callbacks = std::mem::take(&mut self.callbacks).expect("callbacks exist");
+
+        // `tokio::runtime::Handle` is reference counted, so it's efficiently cloneable.
+        let mut reset_instance = Self::new(self.runtime_handle.clone(), callbacks, config).await?;
+
```

Comment:
Why not just use `Option::take`? Don't forget to set back to `Some(config)` and `Some(callbacks)` if `Self::new` returns an error

Reply:
Thanks, good catch! Changed to `take` here: [Use Option::take instead of std::mem::take](https://github.com/NomicFoundation/hardhat/pull/4665/commits/961bdb2af68771c353cab604ea78ea4bb5e5e67c)

And switched back to cloning here, because otherwise we would have to move out the config and the callbacks from `ProviderData::new` on failure: 
[Use clone instead of take in reset](https://github.com/NomicFoundation/hardhat/pull/4665/commits/a1251c7043c35e467b665a0284715735aeec77f0)